List: Add Layout Support

shail-mehta opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Added Layout support and also added Align Wide and full width support to the List block.

Part of : https://github.com/WordPress/gutenberg/issues/43248

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
The block was missing Layout Support and Align Support (Wide and Full Width) Support Features.

## Testing Instructions
- Test the `List block` in both the block editor and site editor.
- Confirm that the wide width, full width and layout settings (including justifications) work correctly, and display correctly on the front.
- Depending on where you place the block, you may need to place it inside a group with the "inner blocks use content width" setting enabled.
- In the Site Editor, open or create the template for single posts. Add a `List block` compare the position of the content inside these blocks, with different settings enabled. It should be possible to align the block so that the text position is the same in `List Block`.

## Screenshots or Screencast <!-- if applicable -->


https://github.com/user-attachments/assets/7f083361-9b3a-4a5a-911c-c76cf53c1b7c

shail-mehta committed:
List: Add Layout Support

shail-mehta committed:
List: Add Layout Support

shail-mehta committed:
fixed lint issue

aaronrobertshaw commented on packages/block-library/src/list/block.json:
```diff
@@ -38,6 +38,8 @@
 	"supports": {
 		"anchor": true,
+		"align": [ "wide", "full" ],
+		"__experimentalLayout": true,
```
I believe `__experimentalLayout` was stabilized to `layout` back in June 2023 via https://github.com/WordPress/gutenberg/pull/51434.

aaronrobertshaw reviewed (commented):
Thanks for the PR @shail-mehta 👍 

Could you please add some further details to the PR description and test instructions? 

I'm not sure exactly what needs to be tested here or the reasoning behind the layout configuration chosen. Also, tweaking the PR title to match what is happening in the code changes is always good (i.e. adding align and layout support).

shail-mehta commented:
Thanks for the Feedback. I will work on this.

shail-mehta committed:
Added Align and Layout Support in List Block

shail-mehta committed:
Fixed Lint Issue

shail-mehta commented:
@aaronrobertshaw I have Used `layout` instead of `__experimentalLayout` and also added testing instructions in description.

aaronrobertshaw commented:
Thanks for the iteration @shail-mehta 👍 

Unfortunately, I don't have the bandwidth to review this in the near future.

To help others test and review this PR the test instructions might need improving along with the title, as mentioned previously.

> - Add List Block in Page / Post.
> - Select and align wide or full width
> - Also Check With layout width in Layout Settings
> - Check Settings in Editor and Frontend Side

For example, check what? What should they change? What's the expected outcome?

shail-mehta commented:
Updated Testing Instructions for `List Block`

carolinan commented:
When the block width is not adjusted, the justification option does not change the position of the list items.
Is there a work around for that or how can we make it easier for users? It will look like the setting is not working.

shail-mehta commented:
Thank you for the review, I will work on this

shail-mehta commented:
@tellthemachines @carolinan Thanks for the feedback 🙇

I have rebased the PR, resolved the conflicts, added block gap support, and fixed the justification issue that occurred when the width was not set.

https://github.com/user-attachments/assets/651a3730-56b1-4bd4-974e-86dbb7933fa4

tellthemachines commented on packages/block-library/src/list/block.json:
```diff
@@ -68,9 +70,11 @@
 				"margin": false,
-				"padding": false
+				"padding": false,
+				"blockGap": false
```
I think we can leave blockGap to display by default

tellthemachines reviewed (commented):
> fixed the justification issue that occurred when the width was not set.

That's not an issue, but how the justification control is meant to work for constrained layout. It justifies the whole block, not its contents. So if the block isn't narrower than its container, it appears to do nothing. In retrospect, "justification" may not have been the best name, because what it does is closer to alignment 😅 but I'm not sure we can change that at this point (at least, not without a migration). 

So let's not try to fix it but instead, it might be better to disable that particular control for this block in block.json:
```
"layout": {
			"allowJustification": false
		},
```

The block-level justification only really makes sense for large container blocks; for an in-content block like List it won't be very useful.

shail-mehta commented:
Thanks for the feedback!

I have disabled justification by setting `allowJustification: false` and also added block gap support by default.

Please let me know if I’ve missed anything.

tellthemachines commented on phpunit/block-bindings-test.php:
```diff
@@ -415,6 +415,26 @@ private function render_list_item_with_source_value( $source_value, $block_conte
 		return trim( $block->render() );
 	}
 
+	/**
```
What's the reason for this change? I'd expect no changes to tests to be needed.

shail-mehta replied:
Thank you for pointing that out. The test change was unintentional, so I’ve reverted it.

tellthemachines replied:
Oh sorry, I see why this change was added now, it was to prevent tests checking List block markup from failing 😅 

There are multiple tests in the `Tests_Block_Bindings` suite checking specific block markup, not just for List block, so it might be best to not change that logic too much. The test updates in this commit sidestep the markup problem, but we can instead just update the test strings to the new markup including the layout classes.

shail-mehta replied:
I’ve added the test data, but I’m not sure if I’m missing anything here.🤔

Would I also need to create a corresponding ticket in Core?

tellthemachines replied:
Yes there are some similar tests using list block markup in core [here](https://github.com/wordpress/wordpress-develop/blob/trunk/tests/phpunit/tests/block-bindings/render.php#L483). The problem is that they will have to be updated simultaneously with the next Gutenberg ref update, otherwise they'll fail because this change doesn't exist in core yet 😅 
I think for now you can just create a draft PR in [wordpress-develop](https://github.com/WordPress/wordpress-develop) and add it in the backport changelog folder to get the check to pass; that PR should probably be linked to a ref update trac ticket once one exists, but that part can be done later.

shail-mehta replied:
Created a draft PR in `wordpress-develop` [#13177](https://github.com/WordPress/wordpress-develop/pull/13177).

Some tests are still failing, and I’m not sure whether the failures are related to this change or not.

andrewserong commented:
Thank you for the work on this PR! After testing it out and thinking about how the layout support works and the effect of its style output on existing themes, I'm not sure it's the right fit for the List block. It's quite nuanced and subtle, but it adversely affects the spacing of nested list items due to how layout zeroes out the top and bottom items:

<img width="1060" height="528" alt="image" src="https://github.com/user-attachments/assets/d4c53ce3-2ec0-4370-9710-12d4e63b7000" />

As such, I'd recommend we skip adding layout support for this block, and let themes use margin and line-height rules where needed to style lists and nested lists.

However, as @tellthemachines pointed out, it'd still be good to add `align` support, that seems really useful. Could we scale this PR back to just adding alignment support?

tellthemachines commented:
> it adversely affects the spacing of nested list items due to how layout zeroes out the top and bottom items

This is a good point, and there's no back-compatible way of adding layout to a block; the styles will always change. 

Also, if we skip adding layout we can revert the test changes and the remaining failing PHP tests should pass again 😅
